db/migrate/20190517161246_add_title_to_ambassador_tasks.rb

```diff
@@ -0,0 +1,9 @@
+class AddTitleToAmbassadorTasks < ActiveRecord::Migration
+  def change
+    change_table :ambassador_tasks do |t|
+      t.column :title, :string, null: false, default: ""
+    end
+
+    add_index :ambassador_tasks, :title, unique: true
```

Comment:
Why add this index? Is it just to make creating/updating faster by validating uniqueness at the db level?

What happens if an ambassador_task is created without a title?

Reply:
My default heuristic for applying data integrity constraints is something like:

1. Add a db-level constraint if one is available
2. Add an app-level constraint if friendlier user-facing errors are needed or if (1) isn't available

There's a presence validator in `AmbassadorTask` that'll keep blank tasks from being created, but if you force it through you can persist one task with a blank title.

```rb
# app/models/ambassador_task.rb L5 (379470e9)

validates :title, presence: true, uniqueness: true
```

Duplicate titles (including duplicate blank ones) will raise an exception if they make it through to the database:

```
2.5.1 (bikeindex)[22] » AmbassadorTask.create
   (0.1ms)  BEGIN
  AmbassadorTask Exists (0.2ms)  SELECT  1 AS one FROM "ambassador_tasks" WHERE "ambassador_tasks"."title" = '' LIMIT 1
   (0.1ms)  ROLLBACK
```

```
2.5.1 (bikeindex)[23] » dupe_task.save(validate: false)
   (0.1ms)  BEGIN
  SQL (0.6ms)  INSERT INTO "ambassador_tasks" ("title", "created_at", "updated_at") VALUES ($1, $2, $3) RETURNING "id"  [["title", "One"], ["created_at", "2019-05-18 01:38:31.581295"], ["updated_at", "2019-05-18 01:38:31.581295"]]
   (0.1ms)  ROLLBACK
ActiveRecord::RecordNotUnique: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_ambassador_tasks_on_title"
DETAIL:  Key (title)=(One) already exists.
: INSERT INTO "ambassador_tasks" ("title", "created_at", "updated_at") VALUES ($1, $2, $3) RETURNING "id"
```